### PR TITLE
Rectification Algorithm Documentation

### PR DESCRIPTION
**Pull request by forman**
Documented the algorithm used in [`rectify_dataset()`](https://xcube.readthedocs.io/en/latest/api.html#xcube.core.resampling.rectify_dataset).

To review, 

```shell
   conda install -c conda-forge sphinx sphinx_rtd_theme sphinx-argparse recommonmark
   pip install sphinx-autodoc-annotation
   sphinx-build docs/source docs/build
```

and open `docs/build/index.html`.

**EDIT**

First merge #956, before merging this one.

### CODE REVIEW

File: docs/source/rectify.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+![Source coordinates](rectify/source-coords.png)
+
+The true geoid surface is fractal, hence there is no defined "best guess" for 
+any point *P(i+u,j+v)* with *0 ≤ u ≤ 1, 0 ≤ v ≤ 1* in between the given coordinates 
+points at *P(i+½, j+½)*. Hence, we use triangulation for its simplicity so 
+that any in-between *P* is found by linear interpolation.
+
+From the coordinates *(P1, P2, P3)* of the first source triangle, the bounding
+box in pixel coordinates in the target image can be exactly determined, 
+because the target grid is regular, *P = x0 + i Δx, y0 - j Δy*, for each
```
[the remark this answers is not in the document]
Comment (forman) [reply docs/source/rectify.md:74]: Will clarify this.

File: docs/source/rectify.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+[grid mapping](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html#grid-mappings-and-projections) 
+into new data variables for a given regular target grid mapping and returns 
+them as a new dataset. 
+
+## Problem Description
+
+The following figure shows a Sentinel-3 OLCI Level-1b scene in its original 
+satellite perspective. In addition to the measured reflectances, the data 
+product also provides the latitude and longitude for each pixel as two 
+individual coordinate images. To the right, the absolute value of the gradient
```
[the remark this answers is not in the document]
Comment (forman) [reply docs/source/rectify.md:17]: Will put formula.

File: docs/source/rectify.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    *V = V3 if v > ½*  
+    *V = V1 else*  
+
+The fractions *u,v* can also be used to perform a bilinear interpolation 
+between the four adjacent measurements pixels:  
+    *VA = V1 + u (V2 - V1)*  
+    *VB = V3 + u (V4 - V3)*  
+    *V = VA + v (VB - VA)*  
+
+with *V1 = V(i,j)*, *V2 = V(i+1,j)*, *V3 = V(i,j+1)*, and *V4 = V(i+1,j+1)*
```
[the remark this answers is not in the document]
Comment (forman) [reply docs/source/rectify.md:111]: Will clarify.

File: docs/source/rectify/olci-input.png
[the remark this answers is not in the document]
Comment (forman) [reply docs/source/rectify/olci-input.png:1]: Will explain them.

File: docs/source/rectify/source-coords.png
[the remark this answers is not in the document]
Comment (forman) [reply docs/source/rectify/source-coords.png:1]: will remove them

**Review by konstntokas — APPROVED** [review]

File: docs/source/rectify.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+product also provides the latitude and longitude for each pixel as two 
+individual coordinate images. To the right, the absolute value of the 
+gradient vectors  of the latitude and longitude  *|(∇ lon)² + (∇ lat)²|* 
+are shown:  
+
+<!-- <img src="rectify/olci-input.png" alt="OLCI L1B Input" style="width:60em;display:block;margin-left:auto;margin-right:auto"/> -->
+![OLCI L1B Input](rectify/olci-input.png)
+
+The given coordinates, latitude and longitude, are _terrain corrected_ with 
+respect to a digital elevation model (DEM) that represents the Earth's geoid.
```
Comment (pont-us) [inline docs/source/rectify.md:25]: I'm confused about the use of "geoid" here and in the rest of the text and diagram. I only know "geoid" as an equipotential gravitational surface which doesn't correspond completely to any set of physical features (as illustrated e.g. [here](https://www.britannica.com/science/geoid) ). But in this text "geoid" seems to be synonymous with "elevation".

File: docs/source/rectify.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+<!-- <img src="rectify/olci-output.png" alt="OLCI L1C Output" style="width:60em;display:block;margin-left:auto;margin-right:auto"/> -->
+![OLCI L1C Output](rectify/olci-output.png)
+
+
+## Algorithm Description
+
+The input to the rectification algorithm is satellite imagery in satellite 
+viewing perspective. In addition, two images - one for each spatial 
+dimension - provide the terrain corrected spatial coordinates.
```
Comment (pont-us) [inline docs/source/rectify.md:49]:
```suggestion
viewing perspective. In addition, two images – one for each spatial 
dimension – provide the terrain corrected spatial coordinates.
```

Typographical pedantry :)

File: docs/source/rectify.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+The expected algorithm input is:
+
+* *N* source measurement images;
+* *2* source coordinate images (one for each spatial dimension) 
+  comprising terrain corrected coordinates *x,y* in units of the CRS
+  for each pixel *i,j*;
+
+The output produced by the algorithm is 
+
+* *N* target measurement images retrieved from source measurement images
```
Comment (pont-us) [inline docs/source/rectify.md:61]:
```suggestion
* *N* target measurement images generated from source measurement images
```

File: docs/source/rectify.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+*(P1, P2, P3)* and *(P2, P4, P3)*, and "paint" fractional source pixel
+coordinates into a new target lookup image. The lookup image can than be used 
+to retrieve the source pixel values for a target image, either by nearest 
+neighbor lookup or by interpolation. 
+
+<!-- <img src="rectify/source-coords.png" alt="Source coordinates" style="width:60em;display:block;margin-left:auto;margin-right:auto"/> -->
+![Source coordinates](rectify/source-coords.png)
+
+The true geoid surface is fractal, hence there is no defined "best guess" for 
+any point *P(i+u,j+v)* with *0 ≤ u ≤ 1, 0 ≤ v ≤ 1* in between the given coordinates
```
Comment (pont-us) [inline docs/source/rectify.md:89]:
```suggestion
any point *P(i+u, j+v)* with *0 ≤ u ≤ 1, 0 ≤ v ≤ 1* in between the given coordinates 
```

File: docs/source/rectify.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+<!-- <img src="rectify/algo-2.png" alt="Algorithm #2" style="width:60em;display:block;margin-left:auto;margin-right:auto"/> -->
+![Algorithm #2](rectify/algo-2.png)
+
+After all source pixels have been processed, the resulting target lookup 
+images containing the fractional source pixel indexes, *i + ½ + u* 
+and *j + ½ + v*, can be used to efficiently map a source measurement image 
+*V* into target measurements image. The pixel values of the source
+measurement image are given as *V1 = V(i,j)*, *V2 = V(i+1,j)*,
+*V3 = V(i,j+1)*, and *V4 = V(i+1,j+1)* here and
```
Comment (pont-us) [inline docs/source/rectify.md:118]:
```suggestion
measurement image are given as *V1 = V(i, j)*, *V2 = V(i+1, j)*,
*V3 = V(i ,j+1)*, and *V4 = V(i+1, j+1)* here and
```

File: docs/source/rectify.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+to determine the pixel value *V* for the target measurement image according 
+to:
+
+*V = V2 if u > ½; V3 if v > ½; V1 else*  
+
+The fractions *u,v* can also be used to perform a linear interpolation 
+between the three adjacent source measurements pixels that formed
+the original triangle:
+
+*V = VA + u (V2 - V1) + u (V3 - V1)*
```
Comment (pont-us) [inline docs/source/rectify.md:135]:
```suggestion
*V = VA + u (V2 − V1) + u (V3 − V1)*  
```

File: docs/source/rectify.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+The fractions *u,v* can also be used to perform a linear interpolation 
+between the three adjacent source measurements pixels that formed
+the original triangle:
+
+*V = VA + u (V2 - V1) + u (V3 - V1)*  
+
+Using bilinear interpolation between the four adjacent source 
+measurements pixels:
+
+*V = VA + v (VB - VA)*
```
Comment (pont-us) [inline docs/source/rectify.md:140]:
```suggestion
*V = VA + v (VB − VA)*  
```

File: docs/source/rectify.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+Using bilinear interpolation between the four adjacent source 
+measurements pixels:
+
+*V = VA + v (VB - VA)*  
+
+with
+
+*VA = V1 + u (V2 - V1)*  
+*VB = V3 + u (V4 - V3)*
```
Comment (pont-us) [inline docs/source/rectify.md:145]:
```suggestion
*VA = V1 + u (V2 − V1)*  
*VB = V3 + u (V4 − V3)* 
```

File: docs/source/rectify.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+coordinate, *i + ½ + u* and *j + ½ + v*, which will be both stored in two 
+target lookup images.
+
+<!-- <img src="rectify/algo-2.png" alt="Algorithm #2" style="width:60em;display:block;margin-left:auto;margin-right:auto"/> -->
+![Algorithm #2](rectify/algo-2.png)
+
+After all source pixels have been processed, the resulting target lookup 
+images containing the fractional source pixel indexes, *i + ½ + u* 
+and *j + ½ + v*, can be used to efficiently map a source measurement image 
+*V* into target measurements image. The pixel values of the source
```
Comment (pont-us) [inline docs/source/rectify.md:116]:
```suggestion
*V* into the target measurement image. The pixel values of the source
```

File: test/core/resampling/test_rectify.py [hunk 13]
```diff
@@ -50,6 +43,10 @@ def test_rectify_2x2_to_default(self):
 
     def test_rectify_2x2_to_7x7(self):
         source_ds = self.new_2x2_dataset_with_irregular_coords()
+        # Add offset to "rad" so it's values do not lie on a plane
```
Comment (pont-us) [inline test/core/resampling/test_rectify.py:46]:
```suggestion
        # Add offset to "rad" so its values do not lie on a plane
```

File: test/core/resampling/test_rectify.py [hunk 14]
```diff
@@ -73,14 +70,104 @@ def test_rectify_2x2_to_7x7(self):
...
+                    [nan, 3.0, 5.0, 5.0, nan, nan, nan],
+                    [nan, nan, 5.0, nan, nan, nan, nan],
+                ],
+                dtype=rad.dtype,
+            ),
+        )
+
+    def test_rectify_2x2_to_7x7_linear_interpol(self):
+        source_ds = self.new_2x2_dataset_with_irregular_coords()
+        # Add offset to "rad" so it's values do not lie on a plane
```
Comment (pont-us) [inline test/core/resampling/test_rectify.py:83]:
```suggestion
        # Add offset to "rad" so its values do not lie on a plane
```

File: test/core/resampling/test_rectify.py [hunk 15]
```diff
@@ -73,14 +70,104 @@ def test_rectify_2x2_to_7x7(self):
...
+                    [_nan_, _nan_, 5.000, _nan_, _nan_, _nan_, _nan_],
                 ],
                 dtype=rad.dtype,
             ),
+            decimal=3,
+        )
+
+    def test_rectify_2x2_to_7x7_bilinear_interpol(self):
+        source_ds = self.new_2x2_dataset_with_irregular_coords()
+        # Add offset to "rad" so it's values do not lie on a plane
```
Comment (pont-us) [inline test/core/resampling/test_rectify.py:123]:
```suggestion
        # Add offset to "rad" so its values do not lie on a plane
```

File: xcube/core/resampling/rectify.py [hunk 16]
```diff
@@ -32,12 +33,14 @@
 from .cf import maybe_encode_grid_mapping
 
 
+_INTERPOLATIONS = {"nearest": 0, "linear": 1, "bilinear": 2}
```
Comment (pont-us) [inline xcube/core/resampling/rectify.py:36]: Why not an enum? I don't insist on it, but it seems like an ideal use case for one.

**Review by pont-us — APPROVED** [review]
Approved, with multiple typographical tweaks and one terminological query.

File: xcube/core/resampling/rectify.py (hunk 16 above)
Comment (forman) [reply xcube/core/resampling/rectify.py:36]: Will do.

File: docs/source/rectify.md (hunk 4 above)
Comment (forman) [reply docs/source/rectify.md:25]: Ok, convinced, I will use just DEM.

**Pull request merged**